### PR TITLE
chore: add leaf preimage handling for update

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

**Comment by rphmeier** [comment]
* **#57** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/57?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/lib.rs [hunk 1]
```diff
@@ -55,7 +55,16 @@ pub struct CommitSpec {
...
     /// should be submitted.
-    pub write_set: Vec<(KeyPath, Option<Value>)>,
+    ///
+    /// The `LeafData` is the preimage of the terminal leaf node encountered when seeking the
+    /// given key in the  current revision of the trie. This may be `Some` even when the key in
+    /// question does not have a prior value, but will always be `Some` when the key in question
+    /// has a prior value.
+    ///
+    /// For example, when seeking key `01011`, there are three possibilities: finding a terminal,
+    /// finding a leaf for `01011`, or finding a leaf for another key, such as `01010`. This is
```
Comment (rphmeier) [inline nomt/src/lib.rs:65]: note: we also need the same for the read set, at least if we are proving.

File: nomt/src/store.rs [hunk 2]
```diff
@@ -95,13 +107,42 @@ pub struct Transaction {
...
 
 impl Transaction {
-    /// Write a value to flat storage as part of the atomic batch.
-    /// Any previous value is overwritten. `None` means delete.
-    pub fn write_value(&mut self, path: KeyPath, value: Option<Vec<u8>>) {
-        let cf = self.shared.db.cf_handle(FLAT_KV_CF).unwrap();
+    /// Write a value to flat storage as well as its current leaf node representation
+    /// as part of the atomic batch. Provide the previous value's hash to clean up any previous leaf
+    /// node.
+    pub fn write_value<H: NodeHasher>(
```
Comment (pepyakin) [inline nomt/src/store.rs:113]: Would be good to note the differences of `Some` vs. `None` for `prev_value` and `value`.

File: nomt/src/lib.rs [hunk 3]
```diff
@@ -56,7 +56,16 @@ pub struct CommitSpec {
     /// All values written during any non-discarded execution paths. A single key may appear
     /// only once - even if a key has intermediate values during execution, only the final change
     /// should be submitted.
-    pub write_set: Vec<(KeyPath, Option<Value>)>,
+    ///
+    /// The `LeafData` is the preimage of the terminal leaf node encountered when seeking the
```
Comment (pepyakin) [inline nomt/src/lib.rs:60]:
```suggestion
    /// The [`LeafData`] is the preimage of the terminal leaf node encountered when seeking the
```

**Review by pepyakin — APPROVED** [review]

**Comment by pepyakin** [comment]
### Merge activity

* **Apr 4, 3:42 AM EDT**: @pepyakin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/57).

**Pull request merged**